language/ruby/language.go

```diff
@@ -65,11 +71,120 @@ func (l *Language) DefaultTestFileSuffix() string {
+	defer cancel()
+
+	if err := injectCoverageTracking(repositoryPath); err != nil {
+		return nil, nil, err
+	}
+
+	coverageFilePath := filepath.Join(repositoryPath, "coverage.json")
+	commandOutput, err := util.CommandWithResult(ctx, logger, &util.Command{
+		Command: []string{
+			tools.SymflowerPath, "test",
+			"--language", "ruby",
+			"--workspace", repositoryPath,
+			"--coverage-file", coverageFilePath,
+		},
+
+		Directory: repositoryPath,
+	})
+	if err != nil {
+		return nil, nil, pkgerrors.WithMessage(pkgerrors.WithStack(err), commandOutput)
+	}
+
+	testsTotal, testsPass, err := parseSymflowerTestOutput(commandOutput)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	testResult = &language.TestResult{
+		TestsTotal: uint(testsTotal),
+		TestsPass:  uint(testsPass),
+	}
+
+	testResult.Coverage, err = language.CoverageObjectCountOfFile(logger, coverageFilePath)
+	if err != nil {
+		return nil, nil, pkgerrors.WithMessage(pkgerrors.WithStack(err), commandOutput)
+	}
 
 	return testResult, problems, nil
 }
 
+var languageRakeTestSummaryRE = regexp.MustCompile(`(\d+) runs, \d+ assertions, (\d+) failures, (\d+) errors, (\d+) skips`)
+
+func parseSymflowerTestOutput(data string) (testsTotal int, testsPass int, err error) {
+	testSummary := languageRakeTestSummaryRE.FindStringSubmatch(data)
+	if len(testSummary) == 0 {
+		return 0, 0, pkgerrors.WithMessage(pkgerrors.WithStack(language.ErrCannotParseTestSummary), data)
+	}
+
+	testsTotal, err = strconv.Atoi(testSummary[1])
+	if err != nil {
+		return 0, 0, pkgerrors.WithStack(err)
+	}
+	testsFailure, err := strconv.Atoi(testSummary[2])
+	if err != nil {
+		return 0, 0, pkgerrors.WithStack(err)
+	}
+	testsErrors, err := strconv.Atoi(testSummary[3])
+	if err != nil {
+		return 0, 0, pkgerrors.WithStack(err)
+	}
+	testsSkips, err := strconv.Atoi(testSummary[4])
+	if err != nil {
+		return 0, 0, pkgerrors.WithStack(err)
+	}
+
+	// The "assertions" cannot be used to calculate the passing tests, cause they also include failed assertions.
+	return testsTotal, testsTotal - testsFailure - testsErrors - testsSkips, nil
+}
+
+// injectCoverageTracking injects our custom coverage tracking logic into a ruby project.
+func injectCoverageTracking(repositoryPath string) error {
+	// Add relative import to our special test initialization to every test file so we can track coverage.
+	testDirectoryEntries, err := os.ReadDir(filepath.Join(repositoryPath, "test"))
+	if err != nil {
+		return pkgerrors.WithMessage(err, "reading test directory")
+	}
+	for _, entry := range testDirectoryEntries {
+		if entry.IsDir() {
+			continue
+		}
+		filePath := filepath.Join(repositoryPath, "test", entry.Name())
+
+		fileContent, err := os.ReadFile(filePath)
+		if err != nil {
+			return pkgerrors.WithStack(err)
+		}
+
+		// Only add the import if it does not exist already.
+		if strings.Contains(string(fileContent), "require_relative \"test_init\"") {
```

Comment:
use ```

Reply:
I have the same thing below but with a `\n`, then the backticks don't work anymore. Would rather keep it consistent.